### DISCUSSION TITLE
How to use selections?

### DISCUSSION THREAD

**Discussion by Eisbrenner** [vaexio/vaex#1986] [Q&A] [answered]
Hi, I like to plot a selection of my data, but I can't figure out how that works with the viz.heatmap.

So the following works
```
df = vaex.open("df.hdf5")
df = df[df.depth > 200]
df = df[df.maxage > 24 * 60 * 60 * 365.25 * 6]
df.viz.heatmap(
    x="i",
    y="j",
    what=np.log(vaex.stat.count() + 1),
    limits="99.7%",
)
```
but I think this runs through the data 3 times? Maybe not, but the docs may tell me not to do it this way? I really have trouble with the docs on this topic.

Anyway, I saw the select method and I think the docs argue that this should be better (?), but I can't get it to work:
```
df = vaex.open("df.hdf5")
df.select(df.maxage > 24 * 60 * 60 * 365.25 * 6, name="maxage")
df.select(df.depth > 200, name="depth")
# fails from here on
df.viz.heatmap(x="i",y="j",selection=["(maxage)&(depth)"],what=np.log(vaex.stat.count() + 1),limits="99.7%")
df.viz.heatmap(x="i",y="j",selection=[True],what=np.log(vaex.stat.count() + 1),limits="99.7%")
df.viz.heatmap(x="i",y="j",selection=True,what=np.log(vaex.stat.count() + 1),limits="99.7%")
df["(maxage)&(depth)"].viz.heatmap(x="i",y="j",what=np.log(vaex.stat.count() + 1),limits="99.7%")
df.viz.heatmap(x="i",y="j",selection=["(df.maxage > 24 * 60 * 60 * 365.25 * 6)&(df.depth > 200)"],what=np.log(vaex.stat.count() + 1),limits="99.7%")
```

the last one seems completely miss guided, it seems in that case maxage and depth are not defined anymore.
```
[KeyError: "Unknown variables or column: '(df.maxage > 24 * 60 * 60 * 365.25 * 6)&(df.depth > 200)'"]()
```

The others mostly throw an error related to some recursion
```
[[... skipping similar frames: NodeVisitor.visit at line 407 (652 times), Expression.dependencies at line 680 (327 times), Expression.variables at line 708 (327 times), Selection.dependencies at line 51 (326 times), NodeTransformer.generic_visit at line 492 (326 times), Expression.variables.<locals>.record at line 697 (326 times), translate at line 603 (326 times), Translator.visit_Name at line 506 (326 times)]]()
# and
[RecursionError: maximum recursion depth exceeded]()
```

So, does one use `df.select` here? Should I not use that?

Best
Ezra

**Comment by JovanVeljanoski** [comment, marked as answer]
Hi, here are some examples, i trust examples would be more valuable than explaining things in words:

```python
# Get example data
import vaex
df = vaex.example()


# option 1: using selections
df.select('x > 5')
df.viz.heatmap('x', 'y', selection='default', f='log1p')

# option 2: using selections with custom names
df.select('x > 5', name='my_sel')
df.viz.heatmap('x', 'y', selection='my_sel', f='log1p')

# option 3: with a string expression directly in the viz method
df.viz.heatmap('x', 'y', selection='y < 0', f='log1p')

# option 4: as a vaex expression directly in the viz method
df.viz.heatmap('x', 'y', selection='y < 0', f='log1p')
```

[There are some advanced plotting examples in the docs.](https://vaex.io/docs/example_advanced_plotting.html)

Hope this helps.
J.

**Reply by Eisbrenner** [reply]
Hej, thanks this helped. Still, some of it is quite surprising. Anyway, so I can't combine selections and 'default' refers to simple the last 'select()' command. Instead, I create combined selections. So for future reference and addition to the docs, here a way on how to use multi condition selections:

```
df = vaex.open(...)
df.select((df.x > a) & (df.y > b) & ...) # the last select becomes the default; alternative is to name it
df.viz.heatmap(...,selection="default",...)
```
or if one, like me, prefers to have the selections individually listed, this may work via string formatting
```
selections = {
    1: "x > a",
    2: "y > b",
    3: "...",
}
df.viz.heatmap(..., selection=f'({") & (".join(selections.values())})', ...)
```

Thanks!

**Reply by JovanVeljanoski** [reply]
You can combine them. `df.select()` has a mode argument. By default it is relace, you can use and/or/xor etc.. Reading the docstrings helps :)

Here is an example

```python
import vaex
df = vaex.example()
df.select('x>0')
df.viz.heatmap('x', 'y', f='log1p', selection='default');
df.select('y>0', mode='and')
df.viz.heatmap('x', 'y', f='log1p', selection='default');
```

**Reply by Eisbrenner** [reply]
I see, yes I didn't see that in https://vaex.io/docs/tutorial.html#Selections-and-filtering, https://vaex.io/docs/tutorial.html#Selections-for-plotting or https://vaex.io/docs/example_advanced_plotting.html and assumed thats it. I am going to spend more time in the API section of the docs from now on, however, that sometimes also is difficult, e.g. for the heatmap in https://vaex.io/docs/api.html#vaex-viz it does not list "selection" under "parameters". (EDIT: found the example to this next part now) It says "selection: shape equals length of selection argument" which I think refers to the fact that it works analogously to what is discussed in https://vaex.io/docs/example_advanced_plotting.html.

Anyway, I'm getting closer, thank you for pointing out to me how "select" actually works! I likely would have missed that for some more time to come.

**Reply by JovanVeljanoski** [reply]
It is quite hard to give an example for everything that vaex can do, since there are so many features.. I guess that is why these discussion board exist. 

Good point about docstrings tho, I will try to fix this as soon as I can.

**Reply by JovanVeljanoski** [reply]
Of course, you do not have to use `df.select()`. If you find it better for your use-case to build a boolean expression some other way with some other logic, and pass that expression to the `selection` argument in any method that supports it, that is fine!

**Reply by Eisbrenner** [reply]
I actually find the `df.select(condition, mode='and/or/xor')` quite neat, so I will try that instead of the string concatenation.

**Reply by JovanVeljanoski** [reply]
The aggregators have a selection argument (e.g. `vaex.agg.mean(.., selection='default')`. 

If you want to apply it on the entire dataframe, then better to just use a filter, i.e. `df[df.x > 5].groupby('y')`

**Reply by Eisbrenner** [reply]
Thanks for answering that too, I thought I had the answer and edited the question but have continued to struggle since then!
